### ISSUE TITLE
Allow for registration of RHEL VM

### ISSUE THREAD

**Issue by hferentschik** [minishift/minishift#137]
This ties into #136, since registration needs to happen prior to provisioning.

**Comment by praveenkumar** [comment]
We need to make sure our register/unregister command also support sub-options because those commands can take lot of different arguments

```
[root@rhel-cdk vagrant]# subscription-manager register -h
Usage: subscription-manager register [OPTIONS]

Register this system to the Customer Portal or another subscription management
service

Options:
  -h, --help            show this help message and exit
  --proxy=PROXY_URL     proxy URL in the form of proxy_hostname:proxy_port
  --proxyuser=PROXY_USER
                        user for HTTP proxy with basic authentication
  --proxypassword=PROXY_PASSWORD
                        password for HTTP proxy with basic authentication
  --username=USERNAME   username to use when authorizing against the server
  --password=PASSWORD   password to use when authorizing against the server
  --serverurl=SERVER_URL
                        server URL in the form of https://hostname:port/prefix
  --insecure            do not check the entitlement server SSL certificate
                        against available certificate authorities
  --baseurl=BASE_URL    base URL for content in form of
                        https://hostname:port/prefix
  --type=UNITTYPE       the type of unit to register, defaults to system
  --name=SYSTEMNAME     name of the system to register, defaults to the
                        hostname
  --consumerid=SYSTEMID
                        the existing system data is pulled from the server
  --org=ORG_KEY         register with one of multiple organizations for the
                        user, using organization key
  --environment=ENVIRONMENT
                        register with a specific environment in the
                        destination org
  --release=RELEASE     set a release version
  --autosubscribe       Deprecated, see --auto-attach
  --auto-attach         automatically attach compatible subscriptions to this
                        system
  --force               register the system even if it is already registered
  --activationkey=ACTIVATION_KEYS
                        activation key to use for registration (can be
                        specified more than once)
  --servicelevel=SERVICE_LEVEL
                        system preference used when subscribing automatically,
                        requires --auto-attach
[root@rhel-cdk vagrant]# subscription-manager unregister -h
Usage: subscription-manager unregister [OPTIONS]

Unregister this system from the Customer Portal or another subscription
management service

Options:
  -h, --help            show this help message and exit
  --proxy=PROXY_URL     proxy URL in the form of proxy_hostname:proxy_port
  --proxyuser=PROXY_USER
                        user for HTTP proxy with basic authentication
  --proxypassword=PROXY_PASSWORD
                        password for HTTP proxy with basic authentication
```

**Comment by hferentschik** [comment]
> We need to make sure our register/unregister command also support sub-options because those commands can take lot of different arguments

Do we? For CDK 2 and the Vagrant setup we only take username and password. Why would we need more here. The minishift CLI has such a huge amount of options already, mirroring all registration commands would be a mess imo. The proxy settings are interesting, but I think they we might need in a global way anyways. Everything else I am not sure about. What use-case do you try to solve?

**Comment by jimmidyson** [comment]
To help with the massive amount of options, you can group options with cobra & even hide common options under a separate command, as OpenShift does with `oc options`.

**Comment by hferentschik** [comment]
Ok, so here is what I think we should do. We need the following pieces (names to be discussed). The idea is to make registration actually orthogonal to provisioning which is a change to our original ideas.

* username and password options as part of the root options (we need the credentials at least during _start_, _stop_ and _delete_
* _Registrar_ interface
  * This interface handles registration, it defines a _register_ and _unregister_ method taking the credentials as parameters
  * We provide for now just a RHEL impl
* _RegistrarDetector_ interface
  * This is very similar to the _ProvisionerDetector_ and will behave almost identical. It also tries to detect the right _Registrar_ based on the release info. In our case it will check for RHEL + variant minishift in which case our _RHELRegistrar_ will be returned
  * The detector will also help us to handle some boundary conditions, for example, username and password are specified, but there is no registrar which can handle the VM. In this case we can abort the command with a nice error message

This aligns very much how provisioning works and keeps the code somewhat generic, preventing RHEL details leaking into the code. All RHEL specifics are encapsulated in the registrar impl. Thoughts?

**Comment by praveenkumar** [comment]
> username and password options as part of the root options (we need the credentials at least during start, stop and delete

Agree, same way we do for `openshift-env` so that if options are set as part of environment that also handled as good way.

> Registrar interface 

We can have the interface it will also easy for us to put some test case around it. I hope.

> RegistrarDetector interface 

I don't think that is a required piece for us. if you see current minishift provisioner it have a provisioner method which already detect if iso is rhel one or not and using same we can perform registration part. 

https://github.com/minishift/minishift/blob/cluster-up/pkg/minishift/provisioner/minishift_provisioner.go#L75-L81

**Comment by praveenkumar** [comment]
With discussion with @hferentschik I got some clear idea about what benefit we get when have Detector Interface for registration. will start works towards it and try to put initial bits in place.

**Comment by hferentschik** [comment]
Merged via pull request pull #222
